core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
```

Comment:
@guozhangwang i thought the lock would be released as soon as the `replicaManager.appendRecords(..)` had returned from writing to the local log. But i guess you are saying the callback could happen immediately? Anyway, makes sense to release the lock before the log append. Thanks